README.md
Bankruptcy
==========

A bankruptcy document parser.

Notes
==========

Bankruptcy is an open source repository to extract content from bankruptcy documents
It was built for use with Courtlistener.com.

Its main goal is to convert bankruptcy documents into readable JSON data.

Further development is intended and all contributors, corrections and additions are welcome.

Background
==========

This was built to help extract content from bankruptcy documents.

Documents
=========

We currently support the following documents in a voluntary petition.

- Bankruptcy Official Form 106 A/B (Property)
- Bankruptcy Official Form 106 D (Secured Creditors)
- Bankruptcy Official Form 106 E/F (Unsecured Creditors)
- Bankruptcy Official Form 106Sum (Statistics)

TODOs
=====

- B 101 (Official Form 101)
- B2030 (Form 2030) (12/15)
- 521.05 (12/1/08)
- Official Form 106C
- Official Form 106G
- Official Form 106H
- Official Form 106I
- Official Form 106J
- Official Form 106Dec
- Official Form 107


Quickstart
==========

    from bankruptcy import extract_all
    results = extract_all(filepath=filepath)

will return a dictionary of the forms (if found) and the contents of the document in the following format.


    {
     'form_106_ab': ...
     'form_106_d': ...
     'form_106_ef': ...
     'form_106_sum': ...
     'info'...
    }  


Some Notes
==========

This tool relies heavily on PDFPlumber.

Somethings to keep in mind this parser has been tested only on digital PDFs
from recent court filings (ie 2018 and earlier).  This parser does not work on scanned
bankruptcy documents and was built and tested on documents from the Pacific Northwest.


Installation
===============

Installing bankruptcy is easy.

::

    pip install bankruptcy

Or install the latest dev version from github

::

    pip install git+https://github.com/freelawproject/bankruptcy.git@master


Testing
==========

    python3 -m unittest test.tests

Future
======

1) Continue to improve and add documents for extraction.
2) Future updates

Deployment
==========

Tag a release with a similar format v1.0.0, update setup.py and push to master.

License
==========

This repository is available under the permissive BSD license, making it easy and safe to incorporate in your own libraries.

Pull and feature requests welcome. Online editing in GitHub is possible (and easy!)


CHANGES.md
# Change Log

## Current

0.0.7 - Fix setup.py

## Past

0.0.6 - Reduce CRUD in requirements, Drop poetry and simplify tests.  Update readme.
0.0.3 - Update to E/F document extraction
0.0.2 - Update setup file to correct missing requirements
0.0.1 - Initial release


test/tests.py
#!/usr/bin/env python3
import os
from glob import iglob
from unittest import TestCase
import sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
sys.path.append(os.path.dirname(SCRIPT_DIR))

from bankruptcy import (
    extract_all,
    extract_official_form_106_a_b,
    extract_official_form_106_d,
    extract_official_form_106_e_f,
    extract_official_form_106_sum,
)


class BankruptcyTest(TestCase):
    def setUp(self) -> None:
        self.root = os.path.dirname(os.path.realpath(__file__))
        self.files = iglob(f"{self.root}/test_assets/*.pdf")
        self.assets = f"{self.root}/test_assets"

    def test_scanned_pdfs(self):
        """Can we gracefully fail on scans?"""

        filepath = f"{self.root}/test_assets/gov.uscourts.orb.507503.11.0.pdf"
        results = extract_all(filepath=filepath)
        self.assertFalse(results)

    def test_can_we_handle_missing_checkboxes(self):
        """Can we handle partially bad PDF can still return content?"""

        filepath = f"{self.assets}/gov.uscourts.ganb.1125040.35.0.pdf"
        results = extract_all(filepath=filepath)
        self.assertEqual(results["form_106_d"]["error"], "Failed to find document.")
        self.assertEqual(len(results["form_106_ef"]["creditors"]), 22)

    def test_offical_form_106_sum(self):
        """Can we extract content from Official Form 106 Sum"""

        filepath = f"{self.root}/test_assets/gov.uscourts.orb.473342.1.0.pdf"
        results = extract_official_form_106_sum(filepath=filepath)
        self.assertTrue(results["7/11/13"])
        self.assertTrue(results["non_consumer_debts"])
        self.assertEqual(results["3_total"], "863,842.00")
        self.assertEqual(results["9g"], "142,500.00")

    def test_official_form_106_e_f(self):
        """Can we extract content for unsecured creditors"""

        filepath = f"{self.root}/test_assets/gov.uscourts.orb.473342.1.0.pdf"
        results = extract_official_form_106_e_f(filepath=filepath)
        self.assertEqual(
            len(results["creditors"]), 19, msg="Failed to extract creditors"
        )
        self.assertEqual(
            results["creditors"][-1]["debtor"],
            ["At least one of the debtors and another"],
        )

    def test_official_form_106_d(self):
        """Can we extract secured creditors from form 106D"""

        filepath = f"{self.root}/test_assets/gov.uscourts.orb.473342.1.0.pdf"
        results = extract_official_form_106_d(filepath=filepath)
        self.assertEqual(
            len(results["creditors"]), 9, msg="Failed to extract creditors"
        )

    def test_official_form_106_a_b(self):
        """Can we extract content from Form 106 A/B"""

        filepath = f"{self.root}/test_assets/gov.uscourts.orb.473342.1.0.pdf"
        results = extract_official_form_106_a_b(filepath=filepath)
        auto = results["cars_land_and_crafts"][3]
        self.assertEqual(
            len(results["cars_land_and_crafts"]),
            4,
            msg="Failed to extract cars_land_and_crafts",
        )
        self.assertEqual(
            auto["make"], "Skido", msg="Failed to extract cars_land_and_crafts"
        )
        self.assertEqual(
            auto["model"], "SM", msg="Failed to extract cars_land_and_crafts"
        )

    def test_all_methods(self):
        """Can we extract content from all four documents?"""

        filepath = f"{self.root}/test_assets/gov.uscourts.orb.473342.1.0.pdf"
        results = extract_all(filepath=filepath)

        self.assertIn("Jr.", results["info"]["debtor_1"])
        self.assertEqual(results["form_106_sum"]["1a"], "325,882.00")
        self.assertEqual(results["form_106_sum"]["9g"], "142,500.00")
        self.assertEqual(
            results["form_106_ef"]["creditors"][0]["name"],
            "Internal Revenue Service",
        )
        self.assertEqual(
            results["form_106_d"]["creditors"][0]["name"], "Ally Financial"
        )
        self.assertEqual(
            results["form_106_ab"]["cars_land_and_crafts"][1]["make"],
            "Chevrolet",
        )


bankruptcy/utils.py
import re
from typing import Any, Dict, Iterator, List, Optional, Tuple, Union

import pdfplumber
from PyPDF2 import PdfFileReader, PdfFileWriter
from PyPDF2.pdf import PageObject

from bankruptcy.fields import (
    make_ab_totals,
    make_car_dict,
    make_creditor_dict,
    make_other_dict,
    make_property_dict,
    make_secured_creditor_dict,
    property_options,
)
from bankruptcy.filters import (
    filter_106_ab_content,
    filter_boxes,
    input_white_text_and_left_side,
    just_text_filter,
    key_filter,
    keys_and_input_text,
    line_filter,
    remove_margin_lines,
)


def get_106_sum_pages(filepath: str) -> List:
    """Find the pages containing Form 106 Sum

    :param filepath:
    :return:
    """
    pages_106 = []
    with pdfplumber.open(filepath) as pdf:
        pages = pdf.pages
        for page in pages:
            if page.extract_text()[-300:].find("Official Form 106Sum") == -1:
                continue
            pages_106.append(page)
    return pages_106


def crop_and_extract(
    page: pdfplumber.pdf.Page,
    line: Dict,
    adjust=False,
    left_shift: int = 0,
    up_shift: int = 20,
) -> str:
    """Extract text content for pdf line if any

    Given a line of a pdf - extracted the text around it.  If adjust
    is True, reduce the cropped area to within the first line (usually above)
    :param page: Page to crop
    :param line: Line to crop around
    :param adjust: Whether to check if another line is inside our crop
    :param left_shift: Leftward crop adjustment
    :param up_shift: Upward crop adjustment
    :return: Content of the section
    """
    bbox = (
        int(line["x0"]) - left_shift,
        int(line["top"]) - up_shift,
        line["x1"],
        line["top"],
    )
    crop = page.crop(bbox)
    if adjust:
        tops = [row["top"] for row in crop.lines if row["top"] != line["top"]]
        if len(tops) > 0:
            crop = page.crop(bbox=(*bbox[:1], tops[-1], *bbox[2:]))
    return crop.filter(keys_and_input_text).extract_text()


def can_we_process_pdf(filepath: str) -> bool:
    """Is this PDF extractable

    :param filepath: Location of the PDF
    :return: Whether it is digital document
    """
    with pdfplumber.open(filepath) as pdf:
        if len(pdf.pages[0].extract_text()) < 100:
            return False
        return True


def convert_pdf(filepath: str, temp_output: str, form: str) -> bool:
    """Find extract and merge forms into one page PDFs

    To simplify extraction we identify the pages in the larger pdf and
    convert them into a single page PDF.

    :param filepath: Location of the pdf to extract
    :param temp_output: temporary filepath
    :param form: Form to find
    :return: Did we succeed
    """
    pages = []
    with pdfplumber.open(filepath) as pdf:
        for page in pdf.pages:
            text = page.extract_text()
            if text[-300:].find(form) == -1:
                continue
            if text[-300:].lower().find("page") == -1:
                continue
            pages.append(page.page_number)

    if not pages:
        return False

    writer = PdfFileWriter()
    with open(filepath, "rb") as pdf_obj:
        pdf = PdfFileReader(pdf_obj)
        first = pages[0] - 1
        last = pages[-1] - 1
        if first == last:
            last = first + 1

        page = pdf.getPage(first)  # We pick the second page here
        height = page.mediaBox.getHeight()
        width = page.mediaBox.getWidth()

        t_page = PageObject.createBlankPage(None, width, height * (last - first + 1))
        length = last - first

        if length == 1:
            writer.addPage(page)
            with open(temp_output, "wb") as file:
                writer.write(file)
        else:
            for pg_number in range(first, last):
                page = pdf.getPage(pg_number)
                t_page.mergeScaledTranslatedPage(
                    page2=page, scale=1, tx=0, ty=height * length, expand=False
                )
                length -= 1
                if not length:
                    last_page = pdf.getPage(last)
                    t_page.mergePage(last_page)
                    writer.addPage(t_page)

            with open(temp_output, "wb") as file:
                writer.write(file)

    print(f"Converted {form}, with {1+last-first} pages, {first} to {last}")
    return True


# ------------ 106 E/F Document
def get_page_and_lines(
    filepath: str,
) -> Tuple[pdfplumber.pdf.Page, List[Dict]]:
    """Get first page and return the lines in the PDF

    :param filepath: Location of the PDF
    :return: PDF page and PDF lines
    """
    with pdfplumber.open(filepath) as pdf:
        only_page = pdf.pages[0]
        all_lines = pdf.pages[0].filter(line_filter).lines
    sorted_lines = sorted(all_lines, key=lambda x: x["top"])
    return only_page, sorted_lines


def parse_unsecured_creditors(page: pdfplumber.pdf.Page, top: int, bottom: int) -> Dict:
    """Extract the information on the unsecured creditor section

    :param page: PDF page
    :param top: Y coordinate of the top of section
    :param bottom: Y coordinate of the bottom of section
    :return: Organized creditor data
    """
    data = []
    crop_one = page.crop((0, max(100, top - 500), page.width, bottom))
    crop = crop_one.crop((0, top, page.width, bottom))
    key = crop.filter(key_filter).extract_text().replace("\n", "")
    boxes = get_checkboxes(crop)
    lines = crop.filter(remove_margin_lines).lines
    for line in sorted(lines, key=lambda x: x["top"]):
        if not data and line["width"] > 20:
            continue
        output = crop_and_extract(crop_one, line, adjust=True, up_shift=100)
        if data or (output is not None and key == output.replace("\n", "")):
            if len(data) == 10 and "2." in key or len(data) == 8 and "4." in key:
                continue
            data.append(output)
    if data:
        return make_creditor_dict(data, boxes, key)
    return {}


def extract_other_creditors_ef(
    page: pdfplumber.pdf.Page, start: Dict, stop: Dict, creditors: List
) -> Dict:
    """Process other creditors to be notified if any

    :param page:Page to crop
    :param start:Y coordinate of the top of the creditor section
    :param stop:Y coordinate of the bottom of the creditor section
    :return: The key, address and acct information
    """

    key_bbox = (start["x0"], start["top"] - 20, start["x1"], start["top"])
    addy_bbox = (0, start["top"] - 20, start["x0"] - 20, stop["top"])
    acct_bbox = (start["x1"] + 150, start["top"] + 20, page.width, stop["top"])

    key = page.crop(key_bbox).filter(just_text_filter).extract_text()
    address = page.crop(addy_bbox).filter(keys_and_input_text).extract_text()
    acct = page.crop(acct_bbox).filter(keys_and_input_text).extract_text()
    for creditor in creditors:
        if creditor["key"] == str(key):
            other_creditors = creditor["other_creditors"]
            other_creditors.append({"address": address, "acct": acct, "key": key})
            creditor["other_creditors"] = other_creditors

    return creditors


# 106 D
def parse_secured_creditors(
    only_page: pdfplumber.pdf.Page, top: int, bottom: int
) -> Dict:
    """Find and extract content from secured creditor portion of 106D

    :param only_page:PDF page
    :param top: Y coordinate for top of section
    :param bottom: Y coordinate of bottom of section
    :return: Organized data of the section
    """
    page = only_page.crop((0, max(100, top - 500), only_page.width, bottom))
    section = page.crop((0, top, only_page.width, bottom))
    key = section.filter(key_filter).extract_text()
    checkboxes = get_checkboxes(section)
    data = []

    for line in sorted(
        section.filter(remove_margin_lines).lines, key=lambda x: x["top"]
    ):
        top = int(line["top"])
        if not data and line["width"] > 20:
            continue
        page_crop = page.crop((line["x0"], top - 200, line["x1"], top))
        tops = [row["top"] for row in page_crop.lines if int(row["top"]) != top]
        if len(tops) > 0:
            if len(data) == 6:
                page_crop = page.crop((line["x0"], tops[-1] - 20, line["x1"], top))
            elif len(data) == 8:
                page_crop = page.crop((line["x0"], top - 50, line["x1"], top))
            else:
                page_crop = page.crop((line["x0"], tops[-1], line["x1"], line["top"]))
        output = page_crop.filter(keys_and_input_text).extract_text()

        if data or key == output:
            data.append(output)

    if data and len(data) > 10:
        return make_secured_creditor_dict(data, checkboxes)
    return {}


def get_checkboxes(crop: pdfplumber.pdf.Page) -> Dict:
    """Find and identify checked checkboxes

    Using multiple tolerances, find checkboxes and identify them by the content
    to the right of the checkbox.

    :param crop: Section of pdf to extract checkboxes from
    :return: Dictionary of selected checkboxes
    """
    results = {}
    # Use multiple tolerances to line up checkboxes on weird PDFs
    for tolerance in [3, 4, 5]:
        filtered_data = crop.filter(filter_boxes).extract_text(y_tolerance=tolerance)
        filtered_data = filtered_data.replace(
            "Type of NONPRIORITY unsecured claim:", ""
        )
        if "[]" not in filtered_data:
            # Checkboxes unreadable
            return {}
        filtered_lines = filtered_data.splitlines()
        checkboxes = [x.replace("  ", " ") for x in filtered_lines if "[" in x]
        query1 = ["debtor"]
        query2 = ["community", "see instructions", "claim relates"]
        query3 = ["No", "Yes"]
        query4 = ["contingent", "unliquidated", "disputed"]
        query5 = [
            "domestic",
            "taxes",
            "death",
            "specify",
            "loans",
            "obligations",
            "pension",
            "including",
            "judgment",
            "statutory",
            "agreement",
        ]

        debtor = [
            box.split(" ", 1)[1].strip()
            for box in checkboxes
            if "√" in box and any(s in box.lower() for s in query1)
        ]
        community = [
            box.split(" ", 1)[1].strip()
            for box in checkboxes
            if "√" in box and any(s in box.lower() for s in query2)
        ]
        offset = [
            box.split(" ", 1)[1].strip()
            for box in checkboxes
            if "√" in box and any(s in box for s in query3)
        ]
        offset = [ans for ans in offset if re.match(r"^(Yes|No)$", ans)]

        info = [
            box.split(" ", 1)[1].strip()
            for box in checkboxes
            if "√" in box and any(s in box.lower() for s in query4)
        ]
        claim_type = [
            box.split(" ", 1)[1].strip()
            for box in checkboxes
            if "√" in box and any(s in box.lower() for s in query5)
        ]

        property_values = [
            box.split(" ", 1)[1].strip()
            for box in checkboxes
            if "√" in box and any(s in box for s in property_options)
        ]
        property_values = [
            s for s in property_options if any(s in box for box in property_values)
        ]

        if claim_type:
            if "Specify" in claim_type[0]:
                claim_type = ["Other. Specify"]
        data = {
            "debtor": debtor,
            "community": community,
            "offset": offset,
            "info": info,
            "claim_type": claim_type,
            "property": property_values,
        }
        if not results:
            results = data
        else:
            datum = [{k: v} for k, v in data.items() if v != []]
            data = {}
            for item in datum:
                data = {**data, **item}

            results = {**results, **data}
    return results


def find_property_sections(
    only_page: pdfplumber.pdf.Page,
) -> Optional[Iterator[Tuple[Union[int, Any], int, Union[int, Any]]]]:
    """Find property sections to iterate over

    Find Sections with white font and identifiers for each section

    :param only_page: PDF page to crop
    :return: None or Iterable sections as top bottom and id
    """
    rows = only_page.filter(input_white_text_and_left_side).extract_words()
    rows = [
        {"top": int(row["top"]), "text": row["text"]}
        for row in rows
        if len(row["text"]) > 2 and row["text"][0] in "P12345" and row["text"][1] == "."
    ]
    if len(rows) == 0:
        return None
    bottoms = [
        int(line["top"])
        for line in only_page.lines
        if line["top"] > rows[0]["top"] and line["width"] > 530
    ][: len(rows)]
    tops = [r["top"] for r in rows]
    keys = [r["text"] for r in rows]
    return zip(tops, bottoms, keys)


# 106 A/B collection
def get_1_to_2_from_a_b(only_page: pdfplumber.pdf.Page) -> List[Dict]:
    """Extract real estate, automobile, jet ski, boats etc, from A/B.

    :param only_page:The PDF page to extract from
    :return: Extracted content
    """
    property_content = []
    sections = find_property_sections(only_page)
    if not sections:
        return property_content

    for top, bottom, key in sections:
        bbox = (0, top, only_page.width, bottom)
        crop = only_page.crop(bbox)
        data = get_all_values_from_crop(crop.lines, only_page)

        if "1." in key:
            section = make_property_dict(key, data)
            checkboxes = get_checkboxes(crop)
            if not checkboxes:
                section["property_interest"] = "Checkbox unreadable"
                section["debtor"] = "Checkbox unreadable"
            else:
                section["property_interest"] = checkboxes["property"]
                section["debtor"] = checkboxes["debtor"]
            property_content.append(section)

        if "3." in key or "4." in key:
            if "3." in key:
                section = make_car_dict(key, data)
            else:
                section = make_other_dict(key, data)

            checkboxes = get_checkboxes(crop)
            if not checkboxes:
                section["debtor"] = "Checkbox unreadable"
            else:
                section["debtor"] = checkboxes["debtor"]
            property_content.append(section)

    return property_content


def get_ab_debtors(rows: List) -> List:
    """Find debtor names to remove from processing

    :param rows: Form content as list of strings
    :return: Debtor names
    """

    debtors = [rows[1]]
    if "[" not in rows[2]:
        match = re.match(r"\d", rows[2][0])
        if not match:
            debtors.append(rows[2])
    return debtors


def clean_ab_data(data: List):
    """Clean data into string or list of strings

    :param data:
    :return:
    """
    matches = re.findall(r".*?\$[\d., $]+", " ".join(data), re.M)
    if matches:
        return matches
    return " ".join(data)


def get_3_to_8_form_a_b(
    page: pdfplumber.pdf.Page,
) -> Tuple[List[Union[Dict[str, Any], Dict[Optional[Any], str]]], list, Optional[dict]]:
    """Parse sections 3 to 8 of 106 A/B property form

    :param page:The pdf page to parse
    :return:Organized property data in the document.
    """
    part = 0
    totals, section, key = None, None, None
    results, data, part_eight = [], [], []

    rows = page.filter(filter_106_ab_content).extract_text().splitlines()
    debtors = get_ab_debtors(rows)

    # Remove debtor rows from lines
    rows = [r for r in rows if r not in debtors]
    for debtor in debtors:
        rows = [r for r in rows if debtor not in r]

    for row in rows[1:]:
        match = re.match(r"Part \d:", row)
        if match:
            part += 1
            continue
        # Extract parts 3 to 7
        if part in [3, 4, 5, 6, 7]:
            match = re.match(r"^\d{1,2}\. ?|^5", row)
            if not match:
                data.append(row)
                continue
            if section == row:
                continue
            if "54. " in row:
                results.append({"54.": row.split(" ")[1]})
            if key:
                data = [d for d in data if "[" not in d]
                if data:
                    if key == "24." and data == ["2"]:
                        data = []
                        continue
                    results.append({key: clean_ab_data(data)})
                data = []
                section = row
            key = row

        if part == 8:
            # Part 8 is the section containing grand totals.
            part_eight.append(row)
            if "63. " in row:  # this is the final row of Part 8
                totals = make_ab_totals(part_eight)

    return results, debtors, totals


def get_all_values_from_crop(lines: List[Dict], page: pdfplumber.pdf.Page) -> List:
    """Find all text inputs for each line in the pdf crop

    :param lines: Possible text inputs
    :param page: PDF to crop
    :return: Text input for each line
    """
    data = []
    for line in lines:
        if line["width"] < 10:
            continue
        output = crop_and_extract(page, line, adjust=True, up_shift=100)
        data.append(output)
    return data


def extract_other_creditors_d(
    page: pdfplumber.pdf.Page, markers: List[Dict], creditors: List
) -> None:
    """Crop and extract address, key and acct # from the PDf

    :param page: PDF page
    :param markers: The top and bottom markers
    :return: Address, key and account information
    """
    adjust = 0 if len(markers) == 5 else 12

    addy_bbox = (
        0,
        markers[0]["top"],
        int(markers[-1]["x1"]) * 0.35,
        markers[-1]["top"],
    )
    key_bbox = (
        markers[-3]["x0"],
        markers[0]["top"] - adjust,
        markers[-3]["x1"],
        markers[-3]["top"],
    )
    acct_bbox = (
        markers[1]["x0"],
        markers[1]["top"] - 12,
        markers[1]["x1"],
        markers[1]["top"],
    )

    address = page.crop(addy_bbox).filter(keys_and_input_text).extract_text()
    key = page.crop(key_bbox).filter(keys_and_input_text).extract_text().strip()
    acct = page.crop(acct_bbox).filter(keys_and_input_text).extract_text()
    for creditor in creditors:
        if creditor["key"] == key:
            other_creditors = creditor["other_creditors"]
            other_creditors.append({"key": key, "address": address, "acct": acct})
            creditor["other_creditors"] = other_creditors
    return creditors


bankruptcy/parser.py
import logging
from tempfile import NamedTemporaryFile
from typing import Any, Dict, List, Union

from bankruptcy.fields import form_106_sum_text_inputs, stats_ef
from bankruptcy.filters import filter_106_sum_boxes, filter_106_sum_lines
from bankruptcy.utils import (
    can_we_process_pdf,
    convert_pdf,
    crop_and_extract,
    extract_other_creditors_d,
    extract_other_creditors_ef,
    get_1_to_2_from_a_b,
    get_3_to_8_form_a_b,
    get_106_sum_pages,
    get_page_and_lines,
    parse_secured_creditors,
    parse_unsecured_creditors,
)


def extract_official_form_106_sum(filepath: str) -> Dict:
    """Extract content from Official Form 106 Sum

    :param filepath: Location of the bankruptcy document
    :return: Data and statistics in Form 106 Sum
    """

    pages = get_106_sum_pages(filepath)
    boxes, inputs = [], []

    if not pages:
        # Pages not found, most likely a scanned PDF document.
        logging.info("Pages not found, may not be a vector pdf")
        return {"error": "Failed to find document."}

    for page in pages:
        checkboxes = page.filter(filter_106_sum_boxes).extract_text()
        if checkboxes:
            boxes.extend(checkboxes.splitlines())

        for line in page.filter(filter_106_sum_lines).lines:
            output = crop_and_extract(page, line, adjust=True, left_shift=5)
            inputs.append(output)

    if not inputs:
        return {}

    # Remove blank inputs
    for item in [6, 7, 10, -1, -2]:
        inputs.pop(item)

    # Analyze Text Inputs
    text_inputs = dict(zip(form_106_sum_text_inputs, inputs))
    checkbox_inputs = {
        "7/11/13": "√" in boxes[2],
        "amended": "√" in boxes[0],
        "consumer_debts": "√" in boxes[3],
        "non_consumer_debts": "√" in boxes[4],
    }
    return {**text_inputs, **checkbox_inputs}


def extract_official_form_106_e_f(
    filepath: str,
) -> Union[Dict[str, str], Dict[str, Union[str, Dict[Any, str], List[dict]]]]:
    """Extract content from 106 EF

    :param filepath: Location of the pdf
    :return: Content of 106 E/F
    """
    section = 0
    results = {}
    creditors, stats, markers = [], [], []

    with NamedTemporaryFile(suffix=".pdf") as file:
        success = convert_pdf(filepath, file.name, "Official Form 106 E/F")
        if not success:
            return {"error": "Failed to find document."}

        page, lines = get_page_and_lines(file.name)

        results["debtor1"] = crop_and_extract(page, lines[0], adjust=True, up_shift=30)
        results["debtor2"] = crop_and_extract(page, lines[1], adjust=True, up_shift=30)
        for line in lines:
            if 498 < line["width"] < 510:
                section += 1

            # Final Section
            if section == 4:
                if line["width"] < 110:
                    output = crop_and_extract(page, line, up_shift=10)
                    stats.append(output)
                    if len(stats) == 10:
                        results["statistics"] = dict(zip(stats_ef, stats))

            # Other sections
            if section == 3:
                if 10 < line["width"] < 20 or line["width"] > 530:
                    markers.append(line)

                if len(markers) == 2:
                    creditors = extract_other_creditors_ef(
                        page,
                        start=markers[0],
                        stop=markers[1],
                        creditors=creditors,
                    )
                    markers = []

            # 1 & 2 Sections
            if line["x0"] < 50 and section in (1, 2):
                # The three marker lines follow these characteristics.
                if len(markers) == 0 and line["width"] > 20:
                    continue
                markers.append(line["top"])
                if len(markers) != 3:
                    continue
                markers.pop(1)

                creditor = parse_unsecured_creditors(
                    page,
                    top=markers.pop(0),
                    bottom=markers.pop(0),
                )
                creditors.append(creditor)

        results["creditors"] = creditors
        return results


def extract_official_form_106_d(
    filepath: str,
) -> Dict[str, Union[List[dict], str]]:
    """Extract content on secured creditors

    :param filepath: Location of the bankruptcy document
    :return: Secured creditor content
    """
    section = 0
    creditors, markers = [], []
    results = {}

    with NamedTemporaryFile(suffix=".pdf") as file:
        success = convert_pdf(filepath, file.name, "Official Form 106D")
        if not success:
            return {"error": "Failed to find document."}
        page, lines = get_page_and_lines(file.name)

    for line in sorted(lines, key=lambda x: x["top"]):
        # if line["width"] < 10:
        #     continue
        if 498 < line["width"] < 510:
            section += 1

        # Other sections
        if section == 2:
            if 5 < line["width"] < 120 or line["width"] > 530:
                markers.append(line)

            if line["width"] > 530:
                # Sometimes there are weird lines. Boxes are not lines
                if len(markers) > 4:
                    creditors = extract_other_creditors_d(page, markers, creditors)
                markers = []

        # Main Sections
        if line["x0"] < 50 and (section == 1):
            if len(markers) == 0:
                if line["width"] > 20:
                    continue
            markers.append(line["top"])
            if len(markers) != 3:
                continue
            markers.pop(1)
            creditor = parse_secured_creditors(
                page, top=markers.pop(0), bottom=markers.pop(0)
            )
            if creditor:
                creditors.append(creditor)

    results["creditors"] = creditors
    return results


def extract_official_form_106_a_b(filepath) -> Dict:
    """Extract property information from official form 106 A/B

    :param filepath:Location of the pdf
    :return: Content from A/B if any
    """
    with NamedTemporaryFile(suffix=".pdf") as file:
        success = convert_pdf(filepath, file.name, "Official Form 106A/B")
        if not success:
            return {"error": "Failed to find document."}

        page, _ = get_page_and_lines(file.name)
        cars_land_and_crafts = get_1_to_2_from_a_b(page)
        results, debtors, totals = get_3_to_8_form_a_b(page)
        return {
            "cars_land_and_crafts": cars_land_and_crafts,
            "debtors": debtors,
            "other_property": results,
            "totals": totals,
        }


def extract_all(filepath: str) -> Union[Dict, bool]:
    """Extract content for all available documents

    :param filepath: Location of the PDF to extract
    :return: Data from bankruptcy document
    """
    processable = can_we_process_pdf(filepath)
    if not processable:
        return False

    form_106_ef = extract_official_form_106_e_f(filepath)
    form_106_sum = extract_official_form_106_sum(filepath)
    form_106_ab = extract_official_form_106_a_b(filepath)
    form_106_d = extract_official_form_106_d(filepath)

    return {
        "info": {
            "debtor_1": form_106_ef["debtor1"],
            "debtor_2": form_106_ef["debtor2"],
        },
        "form_106_ab": form_106_ab,
        "form_106_d": form_106_d,
        "form_106_ef": form_106_ef,
        "form_106_sum": form_106_sum,
    }
